### PR TITLE
GH-266: Delay Rebalance Until Pipeline Drained

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/reactor/reactor-kafka/issues/266
Resolves https://github.com/reactor/reactor-kafka/issues/240

Due to the async nature of reactive workloads, group rebalancing
can occur while records remain from the previous poll(s).
This causes issues with commit failures and record duplication.

Add a configurable maxiimum delay which will be applied when a rebalance
occurs, to wait for records in the pipeline to be processed and let
the rebalance occur at that time.

During the delay, periodically commit any available offsets with the
interval for these commits also being confgurable.

### CODE REVIEW

**Comment by garyrussell** [comment]
Also tested with this (and debug logging):

```java
@SpringBootApplication
public class Rkgh266Application {

	private static final Logger log = LoggerFactory.getLogger(Rkgh266Application.class);

	private final CountDownLatch latch = new CountDownLatch(1);

	public static void main(String[] args) {
		SpringApplication.run(Rkgh266Application.class, args);
	}

	@Bean
	public ApplicationRunner runner1() {
		return args -> {
			SenderOptions<String, String> so = SenderOptions.create(
					Map.of(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092",
							ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
							ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
							ProducerConfig.ACKS_CONFIG, "all",
							ProducerConfig.TRANSACTIONAL_ID_CONFIG, "tx-"));
			KafkaSender<String, String> sender = KafkaSender.create(so);
			Flux<SenderRecord<String, String, String>> recs = Flux.just(pr("foo"), pr("bar"), pr("fail"));
			Flux<Flux<SenderResult<String>>> flux = sender.sendTransactionally(Flux.just(recs));
			flux.subscribe(res -> res.subscribe(result -> {
				System.out.println(result.recordMetadata());
			}));
			Thread.sleep(5000);
		};
	}

	private SenderRecord<String, String, String> pr(String value) {
		return SenderRecord.create("rk266", 0, null, null, value, value + ".corr");
	}

	@Bean
	public ApplicationRunner runner2() {
		return args -> {
			receiver();
			this.latch.await(60, TimeUnit.SECONDS);
			receiver();
		};
	}

	private void receiver() {
		ReceiverOptions<String, String> ro = ReceiverOptions.<String, String> create(
				Map.of(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092",
						ConsumerConfig.GROUP_ID_CONFIG, "rk266",
						ConsumerConfig.MAX_POLL_RECORDS_CONFIG, 3,
						ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"))
				.withKeyDeserializer(new StringDeserializer())
				.withValueDeserializer(new StringDeserializer())
				.addAssignListener(assignments -> log.info("Assigned: " + assignments))
				.commitBatchSize(1)
				.addAssignListener(parts -> this.latch.countDown())
				.subscription(Collections.singletonList("rk266"));
		KafkaReceiver.create(ro)
				.receive()
				.publishOn(Schedulers.single(), 1)
				.doOnNext(record -> {
					System.out.println(record.value() + "@" + record.offset());
					record.receiverOffset().acknowledge();
					try {
						Thread.sleep(3000);
					}
					catch (InterruptedException e) {
						Thread.currentThread().interrupt();
					}
				})
				.subscribe();
	}

	@Bean
	public NewTopic topic() {
		return TopicBuilder.name("rk266").partitions(2).replicas(1).build();
	}

}
```

File: src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java [hunk 1]
```diff
@@ -149,8 +149,29 @@ private void onPartitionsRevoked(Collection<TopicPartition> partitions) {
...
+                long maxDelayRebalance = receiverOptions.maxDelayRebalance().toMillis();
+                if (isActive.get() && maxDelayRebalance > 0) {
+                    long interval = receiverOptions.commitIntervalDuringDelay();
+                    int inPipeline = commitEvent.commitBatch.getInPipeline();
+                    if (inPipeline > 0 || this.awaitingTransaction.get()) {
+                        long end = maxDelayRebalance + System.currentTimeMillis();
+                        do {
+                            try {
+                                log.debug("Rebalancing; waiting for {} records in pipeline", inPipeline);
+                                Thread.sleep(interval);
```
Comment (simonbasle) [inline src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:163]: would it be better to schedule so task `interval` milliseconds in the future, rather than sleep/blocking ? or this is guaranteed to be executed on a `Thread` that can be blocked?

File: reactor-kafka-samples/src/test/java/reactor/kafka/samples/SampleScenariosTest.java [hunk 2]
```diff
@@ -96,8 +97,9 @@ public void kafkaSinkChain() throws Exception {
         sinkChain.runScenario();
         waitForMessages(expected, received1);
         List<Person> expected2 = new ArrayList<>();
-        for (Person p : expected)
+        for (Person p : expected) {
```
Comment (simonbasle) [inline reactor-kafka-samples/src/test/java/reactor/kafka/samples/SampleScenariosTest.java:100]: nitpick: there seems to be unrelated formatting changes in this PR

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java (hunk 1 above)
Comment (bensilvan) [reply src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:163]: From what I know, rebalance happens on the consumption thread, and during rebalance, consumer is not polling messages. I think its better to block the thread while the thread pool continues to process and acknowledge messages. This way we prevent redundant thread pool tasks transfers*

Comment (garyrussell) [reply src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:163]: We must block it to prevent Kafka from completing the rebalance while the pipeline is drained.

I believe it is a blockable thread because `commitSync()` (which is optionally called) blocks.

File: reactor-kafka-samples/src/test/java/reactor/kafka/samples/SampleScenariosTest.java (hunk 2 above)
Comment (simonbasle) [reply reactor-kafka-samples/src/test/java/reactor/kafka/samples/SampleScenariosTest.java:100]: as far as I can see, no meaningful change but formatting in that file (and marking a field above as `final`), so I would lean towards removing the formatting just in this file

Comment (simonbasle) [reply reactor-kafka-samples/src/test/java/reactor/kafka/samples/SampleScenariosTest.java:100]: nevermind that, there's a method call added to a method call chain below in `kafkaTransform`. I don't mind terribly, just a habit of avoiding large formatting changes in unmodified lines in other reactor projects.

**Review by simonbasle — APPROVED** [review]
